File: src/Atlasd/Battlenet/Protocols/Game/GameAd.cs
using Atlasd.Daemon;
using System;
using System.Collections.Generic;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game
{
    class GameAd
    {
        public enum GameTypes : UInt16
        {
            Melee = 0x02,
            FreeForAll = 0x03,
            FFA = FreeForAll,
            OneVsOne = 0x04,
            CaptureTheFlag = 0x05,
            CTF = 0x05,
            Greed = 0x06,
            Slaughter = 0x07,
            SuddenDeath = 0x08,
            Ladder = 0x09,
            IronManLadder = 0x10,
            UseMapSettings = 0x0A,
            UMS = UseMapSettings,
            TeamMelee = 0x0B,
            TeamFreeForAll = 0x0C,
            TeamFFA = TeamFreeForAll,
            TeamCaptureTheFlag = 0x0D,
            TeamCTF = TeamCaptureTheFlag,
            TopVsBottom = 0x0F,
            PGL = 0x20,
        };

        public enum LadderTypes : UInt32
        {
            None = 0x00,
            Ladder = 0x01,
            IronManLadder = 0x03,
        };

        [Flags]
        public enum StateFlags : UInt32
        {
            None = 0x00,
            Private = 0x01,
            Full = 0x02,
            HasPlayers = 0x04,
            InProgress = 0x08,
            DisconnectIsLoss = 0x10,
            Replay = 0x80,
        };

        public StateFlags ActiveStateFlags { get; private set; }
        public UInt32 ElapsedTime { get; private set; }
        public IList<GameState> Clients { get; private set; }
        public UInt32 GamePort { get; private set; }
        public GameTypes GameType { get; private set; }
        public UInt32 GameVersion { get; private set; }
        public LocaleInfo Locale { get; private set; }
        public byte[] Name { get; private set; }
        public GameState Owner { get => Clients != null && Clients.Count > 0 ? Clients[0] : null; }
        public byte[] Password { get; private set; }
        public Product.ProductCode Product { get; private set; }
        public byte[] Statstring { get; private set; }
        public ushort SubGameType { get; private set; }

        public GameAd(GameState client, byte[] name, byte[] password, byte[] statstring, UInt32 gamePort, GameTypes gameType, ushort subGameType, UInt32 gameVersion)
        {
            ActiveStateFlags = StateFlags.None;
            Clients = new List<GameState>(){client};
            ElapsedTime = 0;
            GamePort = gamePort;
            GameType = gameType;
            GameVersion = gameVersion;
            Name = name;
            Locale = client.Locale;
            Password = password;
            Product = client.Product;
            Statstring = statstring;
            SubGameType = subGameType;
        }

        public bool AddClient(GameState client)
        {
            lock (Clients)
            {
                if (Clients.Contains(client)) return false;
                Clients.Add(client);
            }
            return true;
        }

        public bool HasClient(GameState client)
        {
            lock (Clients) return Clients.Contains(client);
        }

        public bool RemoveClient(GameState client)
        {
            bool removed = false;
            lock (Clients) removed = Clients.Remove(client);
            return removed;
        }

        public void SetActiveStateFlags(StateFlags newFlags)
        {
            ActiveStateFlags = newFlags;
        }

        public void SetElapsedTime(UInt32 newElapsedTime)
        {
            ElapsedTime = newElapsedTime;
        }

        public void SetGameType(GameTypes newGameType)
        {
            GameType = newGameType;
        }

        public void SetGameVersion(UInt32 newGameVersion)
        {
            GameVersion = newGameVersion;
        }

        public void SetName(byte[] newName)
        {
            Name = newName;
        }

        public void SetPassword(byte[] newPassword)
        {
            Password = newPassword;
        }

        public void SetPort(UInt32 newPort)
        {
            GamePort = newPort;
        }

        public void SetStatstring(byte[] newStatstring)
        {
            Statstring = newStatstring;
        }

        public void SetSubGameType(byte newSubGameType)
        {
            SubGameType = newSubGameType;
        }
    };
}

File: src/AtlasV/Battlenet/Protocols/Game/Other/GameAd.vb
Imports System

Namespace AtlasV.Battlenet.Protocols.Game
    Class GameAd
        Public Enum GameTypes As UInt16
            Melee = &H2
            FreeForAll = &H3
            FFA = FreeForAll
            OneVsOne = &H4
            CaptureTheFlag = &H5
            CTF = &H5
            Greed = &H6
            Slaughter = &H7
            SuddenDeath = &H8
            Ladder = &H9
            IronManLadder = &H10
            UseMapSettings = &HA
            UMS = UseMapSettings
            TeamMelee = &HB
            TeamFreeForAll = &HC
            TeamFFA = TeamFreeForAll
            TeamCaptureTheFlag = &HD
            TeamCTF = TeamCaptureTheFlag
            TopVsBottom = &HF
            PGL = &H20
        End Enum

        Public Enum LadderTypes As UInt32
            None = &H0
            Ladder = &H1
            IronManLadder = &H3
        End Enum

        <Flags>
        Public Enum StateFlags As UInt32
            None = &H0
            [Private] = &H1
            Full = &H2
            HasPlayers = &H4
            InProgress = &H8
            DisconnectIsLoss = &H10
            Replay = &H80
        End Enum

        Public Property ActiveStateFlags As StateFlags
        Public Property ElapsedTime As UInt32
        Public Property Client As GameState
        Public Property GamePort As UInt32
        Public Property GameType As GameTypes
        Public Property GameVersion As UInt32
        Public Property Name As Byte()
        Public Property Password As Byte()
        Public Property Statstring As Byte()
        Public Property SubGameType As UShort

        Public Sub New(ByVal varClient As GameState,
                       ByVal varName As Byte(),
                       ByVal varPassword As Byte(),
                       ByVal varStatstring As Byte(),
                       ByVal varGamePort As UInt32,
                       ByVal varGameType As GameTypes,
                       ByVal varSubGameType As UShort,
                       ByVal varGameVersion As UInt32)
            ActiveStateFlags = StateFlags.None
            Client = varClient
            ElapsedTime = 0
            GamePort = varGamePort
            GameType = varGameType
            GameVersion = varGameVersion
            Name = varName
            Password = varPassword
            Statstring = varStatstring
            SubGameType = varSubGameType
        End Sub

        Public Sub SetActiveStateFlags(ByVal newFlags As StateFlags)
            ActiveStateFlags = newFlags
        End Sub

        Public Sub SetElapsedTime(ByVal newElapsedTime As UInt32)
            ElapsedTime = newElapsedTime
        End Sub

        Public Sub SetGameType(ByVal newGameType As GameTypes)
            GameType = newGameType
        End Sub

        Public Sub SetGameVersion(ByVal newGameVersion As UInt32)
            GameVersion = newGameVersion
        End Sub

        Public Sub SetName(ByVal newName As Byte())
            Name = newName
        End Sub

        Public Sub SetPassword(ByVal newPassword As Byte())
            Password = newPassword
        End Sub

        Public Sub SetPort(ByVal newPort As UInt32)
            GamePort = newPort
        End Sub

        Public Sub SetStatstring(ByVal newStatstring As Byte())
            Statstring = newStatstring
        End Sub

        Public Sub SetSubGameType(ByVal newSubGameType As Byte)
            SubGameType = newSubGameType
        End Sub
    End Class
End Namespace